org.eclipse.xtext.ui.shared/src/org/eclipse/xtext/ui/shared/internal/ListenerRegistrar.java

```diff
@@ -41,9 +46,25 @@ public void initialize() {
 		// after the fact, the project-open event is already done. Thus no build is triggered for the newly
 		// opened project and the index state is corrupt.
 		// thus we trigger a recovery build here
-		if (builderState.isEmpty()) {
-			buildManager.scheduleBuildIfNecessary(Arrays.asList(workspace.getRoot().getProjects()), IBuildFlag.RECOVERY_BUILD);
+		class RecoveryBuildTrigger extends WorkspaceJob {
+
+			public RecoveryBuildTrigger() {
+				super("Schedule Xtext recovery build on start-up");
+			}
+
+			@Override
+			public IStatus runInWorkspace(IProgressMonitor monitor) throws CoreException {
+				if (monitor.isCanceled()) {
+					return Status.CANCEL_STATUS;
+				}
+				if (builderState.isEmpty()) {
+					buildManager.scheduleBuildIfNecessary(Arrays.asList(workspace.getRoot().getProjects()), IBuildFlag.RECOVERY_BUILD);
+				}
+				return Status.OK_STATUS;
+			}
 		}
+		RecoveryBuildTrigger recoveryBuildTrigger = new RecoveryBuildTrigger();
```

Comment:
I would add a comment why this is done in the job.

The commit message could also have some more context

Reply:
Andrey, Christian, Sebastian, please review this and +1 when you are done, as in a gerrit review.